src/server/server_family.cc

```diff
@@ -2290,40 +2370,41 @@ void ServerFamily::ReplicaOfInternal(string_view host, string_view port_sv, Conn
     return;
   }
 
+  OpResult<ReplicaOfArgs> replicaof_args =
+      ReplicaOfArgs::FromParams(host, port, slot_start, slot_end);
+  if (!replicaof_args) {
+    return cntx->SendError(replicaof_args.status());
+  }
+
   // If NO ONE was supplied, just stop the current replica (if it exists)
-  if (IsReplicatingNoOne(host, port_sv)) {
+  if (replicaof_args->no_one) {
     if (!ServerState::tlocal()->is_master) {
       CHECK(replica_);
 
       SetMasterFlagOnAllThreads(true);  // Flip flag before clearing replica
       replica_->Stop();
       replica_.reset();
+
+      for (auto& replica : cluster_replicas_) {
+        replica->Stop();
+      }
+      cluster_replicas_.clear();
     }
 
-    CHECK(service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE).first ==
-          GlobalState::ACTIVE)
+    CHECK(service_.SwitchState(GlobalState::LOADING, GlobalState::ACTIVE) == GlobalState::ACTIVE)
         << "Server is set to replica no one, yet state is not active!";
 
     return cntx->SendOk();
   }
 
-  uint32_t port;
-  if (!absl::SimpleAtoi(port_sv, &port) || port < 1 || port > 65535) {
-    cntx->SendError(kInvalidIntErr);
-    return;
-  }
-
-  // First, switch into the loading state
-  if (auto new_state = service_.SwitchState(GlobalState::ACTIVE, GlobalState::LOADING);
-      new_state.first != GlobalState::LOADING) {
-    LOG(WARNING) << GlobalStateName(new_state.first) << " in progress, ignored";
-    cntx->SendError("Invalid state");
-    return;
-  }
-
   // If any replication is in progress, stop it, cancellation should kick in immediately
   if (replica_)
     replica_->Stop();
+  // Stop all cluster replication.
+  for (auto& replica : cluster_replicas_) {
+    replica->Stop();
+  }
+  cluster_replicas_.clear();
```

Comment:
This can be wrapped in a small utility function
btw don't we need to call `replica_.reset()` like above?